### PR TITLE
metrics: storage: measure cpu and memory as well as bandwidth

### PR DESCRIPTION
**Pull request by grahamwhaley**
Whilst running the storage (fio) tests, also measure the CPU
consuption and the memory footprint being used.
To do this we re-structure the test to set up the initial
test environment (get the container up and then init the
test file - which takes longer than the test itself), and then
we can know exactly when we are running the actual test.

Fixes: #630

Signed-off-by: Graham whaley <graham.whaley@intel.com>

### CODE REVIEW

File: metrics/storage/fio_job.sh [hunk 1]
```diff
@@ -72,24 +77,49 @@ EOF
...
+# this job will run inside of the container.
+# This job runs the actual test
+function create_fio_run_job()
+{
+	FIO_JOB="$(cat << EOF
+	#!/bin/bash
+	# This fio job will run inside of
+	# the container.
+
+	fio --ioengine="$ENGINE" \
```
Comment (jodh-intel) [inline metrics/storage/fio_job.sh]: We might not care (and I see there are already existing functions similar to this), but you could make that `exec fio` to avoid creating a 2nd process. In fact, why don't we just call `fio` directly? I've just noticed that `$FIO_JOB` itself is called via `bash -c` so we have 3 processes I think?

File: metrics/storage/fio_job.sh [hunk 2]
```diff
@@ -72,24 +77,49 @@ EOF
 
 # Create the job configuration for FIO
 # this job will run inside of the container.
-function create_fio_job()
+# This job sets up the image - creates the test file
+function create_fio_prep_job()
 {
 	FIO_JOB="$(cat << EOF
```
Comment (grahamwhaley) [inline metrics/storage/fio_job.sh]: Just found a couple of uses of `FIO_JOB`, which got deprecated when I split it into RUN and PREP phases ... think I got bit by a rebase.... new push coming.

File: metrics/storage/fio_job.sh [hunk 3]
```diff
@@ -125,7 +155,7 @@ function main()
 {
 	cmds=("bc" "awk")
```
Comment (MarioCarrilloA) [inline metrics/storage/fio_job.sh]: we can ask for `smem` tool

File: metrics/storage/fio_job.sh [hunk 4]
```diff
@@ -161,14 +194,42 @@ function main()
...
+	docker exec -ti ${CONTAINER_ID} bash -c "$FIO_PREP_JOB"
+
+	# Kick off background tasks to measure the CPU, PSS and RSS
+	cpu_temp=$(mktemp fio_job_cpu.XXX)
+	pss_temp=$(mktemp fio_job_pss.XXX)
+	rss_temp=$(mktemp fio_job_rss.XXX)
+
+	# We know the test will run for 30s (unless somebody changed it...)
+	# So spread the measures out over the test run
+	(sleep 10; ps --no-headers -o %cpu -C $(basename ${QEMU_PATH}) > ${cpu_temp})&
```
Comment (MarioCarrilloA) [inline metrics/storage/fio_job.sh]: I made a test and it shows and error using `runc`, it seems that the test is looking for a `qemu` instance, maybe we could check the runtime and in the case of `runc` we can measure the `fio` process running in host

**Comment by GabyCT** [comment]
@grahamwhaley while running the test I am getting using CC

===== starting test [storage IO random write bs 16k] =====
command: docker: yes
command: bc: yes
command: awk: yes
Units: KB
389936.00 KB/s
ERROR: result argument not supplied

File: metrics/storage/fio_job.sh (hunk 4 above)
Comment (grahamwhaley) [reply metrics/storage/fio_job.sh]: Yeah. I knew this was a gap when I did the PR - I was going to ask @GabyCT what 'process' she measured when not running a VM runtime - and I guess the logical answer is the `fio` process - so, sure, let me make that modification...

**Comment by grahamwhaley** [comment]
Hi @GabyCT Yeah, I found an issue where the write test seems to not honour the `--runtime=30` request if it, probably, manages to cover the whole file - so the test quits before the measurements take place, and then we end up with an empty result, which baulks :-(
I have made a modification to take the samples at 3/4/5s - my write test was completing in 13s for reference. What a pain ;-)
I have some other mods coming as well.
Oh, and another note - if we use the `-x magic args`, they include a `--ramp_time=60 --runtime=300` - which messes the test as well - in that case I set the sample sleep times to 65/66/67 to get past the ramp_time.

So, now I have give us the `-x` argument, that sort of can mess with this PRs intention - I think 'such is life', as we can never make the script smart enough to handle everything...

File: metrics/storage/fio_job.sh (hunk 4 above)
Comment (mcastelino) [reply metrics/storage/fio_job.sh]: @GabyCT can you confirm which process is used in the case of runc?

**Comment by GabyCT** [comment]
@grahamwhaley and @mcastelino for runc it is the fio process

File: metrics/storage/fio_job.sh (hunk 1 above)
Comment (grahamwhaley) [reply metrics/storage/fio_job.sh]: Heh - we probably don't care much, but I took out the `#!/bin/bash`, and the world seems fine...
I've not injected the `exec`, as I've not seen that as a common use case on a docker command line.
Mostly I'd like to keep the script readable - I think that is how the original layout came about for those commands.

**Comment by grahamwhaley** [comment]
Updated with feedback. Fixed up some niggles. Rebased and re-pushed.

**Comment by GabyCT** [comment]
lgtm

**Comment by GabyCT** [comment]
@grahamwhaley could you please check this?, I want to merge this but it has a conflict. Thanks

**Comment by grahamwhaley** [comment]
rebased, resolved the merge issue with my other 'rm the fio containers' that already merged, and re-pushed...
